src/back/wgsl/writer.rs

```diff
@@ -122,23 +213,395 @@ impl<W: Write> Writer<W> {
+                    if global {
+                        // First write only constant name
+                        write!(self.out, "let {}: ", name)?;
+                        // Next write constant type and value
+                        match *value {
+                            crate::ScalarValue::Sint(value) => {
+                                write!(self.out, "i32 = {}", value)?;
+                            }
+                            crate::ScalarValue::Uint(value) => {
+                                write!(self.out, "u32 = {}", value)?;
+                            }
+                            crate::ScalarValue::Float(value) => {
+                                write!(self.out, "f32 = {}", value)?;
+                            }
+                            crate::ScalarValue::Bool(value) => {
+                                write!(self.out, "bool = {}", value)?;
+                            }
+                        };
+                        // End with semicolon and extra newline for readability
+                        writeln!(self.out, ";")?;
+                        writeln!(self.out)?;
+                    } else {
+                        write!(self.out, "{}", name)?;
+                    }
+                } else {
+                    match *value {
+                        crate::ScalarValue::Sint(value) => {
+                            write!(self.out, "{}", value)?;
+                        }
+                        crate::ScalarValue::Uint(value) => {
+                            write!(self.out, "{}", value)?;
+                        }
+                        // TODO: fix float
+                        crate::ScalarValue::Float(value) => {
+                            write!(self.out, "{:.1}", value)?;
+                        }
+                        crate::ScalarValue::Bool(value) => {
+                            write!(self.out, "{}", value)?;
+                        }
+                    };
+                }
+            }
+            _ => todo!("write_constant {:?}", constant.inner),
+        }
+
+        Ok(())
+    }
+
     pub fn finish(self) -> W {
         self.out
     }
 }
+
+fn built_in_str(built_in: crate::BuiltIn) -> &'static str {
+    use crate::BuiltIn;
+    match built_in {
+        BuiltIn::VertexIndex => "vertex_index",
+        BuiltIn::InstanceIndex => "instance_index",
+        BuiltIn::Position => "position",
+        BuiltIn::FrontFacing => "front_facing",
+        BuiltIn::FragDepth => "frag_depth",
+        BuiltIn::LocalInvocationId => "local_invocation_id",
+        BuiltIn::LocalInvocationIndex => "local_invocation_index",
+        BuiltIn::GlobalInvocationId => "global_invocation_id",
+        BuiltIn::WorkGroupId => "workgroup_id",
+        BuiltIn::WorkGroupSize => "workgroup_size",
+        BuiltIn::SampleIndex => "sample_index",
+        BuiltIn::SampleMask => "sample_mask",
+        _ => todo!("built_in_str {:?}", built_in),
```

Comment:
not all builtins are going to be mappable to WGSL, so we should make this return an `Error`

Reply:
I already face it when I tried to convert `spv` to `wgsl`. In `spv` we have `gl_PointSize`, for example. And I can't find anything about it in WGSL Spec.  What should we do in this case?

Reply:
Like I said, we should return an error variant in this case